### ISSUE TITLE
Injecting input with kernel 5.7.11 does not work.

### ISSUE THREAD

**Issue by StanGenchev** [gvalkov/python-evdev#129]
I'm running Fedora 32 and this morning, after updating my system, I found that injecting input no longer works. I can solve the issue by booting with an older kernel such as 5.7.9.

Finding and listing devices works, creating a device with UInput works but when I use the write function nothing happens. No error output, no messages, no key events registered.

**Comment by LinusCDE** [comment]
Hi,

initially it seems not the be related to changes in the official kernel specificially. I also have 5.7.9 which was built on Jul 16 (Full version: 5.7.9-1-MANJARO #1 SMP PREEMPT Thu Jul 16 08:20:05 UTC 2020 x86_64 GNU/Linux).
I tested one of my projects that also injects some events using evdev (checked it to be 1.3.0) and it works fine. I also doubt that it is because of some fedora specific kernel compilation parameters.

Maybe it's because of wayland on Fedora.
The event system in linux is a linux standard. There are drivers that fetch the input for use on the desktop which sometimes don't recognize a device properly.

Device recognition is often done by checking what types of buttons you declare to use. Some apps also check for some variation of a name for niche cases (e.g. Krita accepts pens only when they end with "Pen").

I personally had not luck getting my Pen emulation working on libinput at the time (more modern X input driver that sometimes completely replaces xorg-input-evdev). But this seemed to be a driver issue of libinput.

Maybe also try xev for debugging. It is probably geared towards Xorg instead of wayland, but may still help.
Also try looking for what capability/category libinput give's your device with `libinput list-devices`.

To be sure your events get injected at all, I recommend the tool `evtest`. It should definitely show up there and log all injected events.

**Comment by LinusCDE** [comment]
This diagram shows how I understand the system so far (so don't take it by its word):

![graphic](https://user-images.githubusercontent.com/22298664/89336523-d2036680-d699-11ea-9b74-9d0706145036.png)

(The arrows are meant to show in what direction different kinds of input events go.)

As I understand it, xorg-input-evdev (or xf86-input-evdev however you name it) is like the "legacy" driver that translates input events from the kernel (evdev) (device that has capability of codes EV_KEY, KEY_A, ...) to xorgs internal representation of devices (Mouse, Keyboard, ...).

libinput seems to be some kind of update in that regard that also supports wayland. Since the driver is "newer", it sometimes doesn't translate devices correctly to xorg/wayland as far as I understand it. Or it has just harder requirements than evdev.

Xorg usually uses a combination of those drivers. You can actually select in its configuration, which drivers does which "Product". Simularly to how you might assign your display to your graphics driver to e.g. use the nvidia driver on xorg.

With `evtest`, you can test the kernel part. So whether python-evdev actually sends those events to the kernel (which drivers can pic up again).
With the `libinput list-devices` command you can show what libinput sees and gives to wayland.

**Comment by StanGenchev** [comment]
Ok, not 100% sure, but it seams the problem is fixed and it was my fault in the first place.

I had installed evdev globally (which is something I shouldn't do, I know) and the last time I updated my system, it installed (or upgraded, although I don't remember installing it) python3-evdev-1.1.2-7.fc32. That got mixed with the pip install and despite pip showing version 1.3.0, it was a mix of the two.
I removed evdev completely, installed the fedora package version, tested it and I can confirm it works. Then I installed the latest pip version, and that seams to be working as well.

I did this on my office computer and I will test it later today on my personal one as I had the same issue there as well.
If this truly was just the two versions mixing I will close this issue.

**Comment by StanGenchev** [comment]
Turns out it may not be my fault.

On my personal computer the fedora package has never been installed and the pip version is inside a venv.

After running the python script, the device is created and can be seen by ```evtest``` as "py-evdev-uinput" and it does register the EV_KEY but the system doesn't respond.
Running ```libinput list-devices``` reveals that "py-evdev-uinput" is not seen, which explanes why the system doesn't respond.
I have checked for other possibilities such as not being in the ```input``` group and not running as root but I could not locate the source of the issue. Also on my personal system, it does not matter which kernel version I boot into, it never works (but it did work prior to upgrading to 5.7.11).
My work PC and my personal one are both updated and running the same software.
This is a weird one and I'm not sure what I need to do in order to find the issue. Any suggestions?

Again, my stack is:
Fedora 32 (nothing custom, I like clean/stock systems)
Linux kernel 5.7.11
libinput 1.15.902
python-evdev 1.3.0

**Comment by LinusCDE** [comment]
> I had installed evdev globally (which is something I shouldn't do, I know)

In that case you're probably already better than me.

>After running the python script, the device is created and can be seen by evtest as "py-evdev-uinput" and it does register the EV_KEY but the system doesn't respond.

Just for clarity: Did you register also all the possible KEY_*, BTN_*, ... codes for EV_KEY? Only having EV_KEY would probably not help libinput at all since vastly different equipment can use that.

> py-evdev-uinput

That is the default name. You can change that name, but in this case it might not make sense.

---

Also what kind device are you trying to emulate? For plain keyboard or mouse, there are far simpler and cross-platform libraries available (though delving into the nitty gritty is always a cool thing to do).

**Comment by StanGenchev** [comment]
Did more digging and turns out this is definitely an evdev+libinput bug but let me answer your questions first.

> Just for clarity: Did you register also all the possible KEY__, BTN__, ... codes for EV_KEY? Only having EV_KEY would probably not help libinput at all since vastly different equipment can use that.

Yes, it did register. Here is the output:
```
Event: time 1596997549.252782, type 1 (EV_KEY), code 125 (KEY_LEFTMETA), value 1
Event: time 1596997549.252782, type 1 (EV_KEY), code 125 (KEY_LEFTMETA), value 0
```

> Also what kind device are you trying to emulate? For plain keyboard or mouse, there are far simpler and cross-platform libraries available (though delving into the nitty gritty is always a cool thing to do).

Not emulate but translate. I have a Lenovo Wireless USB Touchpad which is not a touchpad but a keyboard/mouse combo (yes, you read that correctly). It is a touchpad physically but identifies as a keyboard computer-side, so for all the 2, 3, 4 and 5 finger gestures, it simply sends keyboard combinations/shortcuts. The problem is that these shortcuts are for Windows 8 so they do not work for Linux. Remapping globally would mess with my actual keyboard and remapping a particular device using ```udev hwdb``` did work for a short while for some of my use cases but not only did that require me to mess with system files, it is also an unofficial/undocumented API so it keeps changing/breaking. I wanted a simple and stable way to control my device and ```python-evdev``` has been working for me for about 2 years now.

After some experimentation, I found out that if I use the ```from_device``` function like this
```
key_input = UInput.from_device(<evdev.InputDevice>)
```
everything works flawlessly, with any version of libinput. Before that I used
```
key_input = UInput()
```
directly, as it did work, but since version 1.15.902 (and 1.16.0) of libinput, that no longer works. I downgraded to version 1.15.4 and I have confirmed this. I tried to add name, vendor, product, bustype and version to the ```UInput()``` constructor but it did not work.
I also tested with different kernel versions and that did not make a difference. libinput was apparently updated along with the kernel, hence my confusion about thinking this is a kernel related bug.

The problem is definitely in the newer versions of libinput when combined with evdev, I just don't know where exactly. This may be a libinput-only bug, it may be evdev-only bug or a bug in both. I just don't have the time and knowledge (yet) to figure it out so any help is appreciated.

Not being able to use ```UInput()``` without inheriting properties from a real device is a problem as you may not always have a physical device to copy from.

**Comment by ssieb** [comment]
> > Just for clarity: Did you register also all the possible KEY__, BTN__, ... codes for EV_KEY? Only having EV_KEY would probably not help libinput at all since vastly different equipment can use that.
> 
> Yes, it did register. Here is the output:
> 
> ```
> Event: time 1596997549.252782, type 1 (EV_KEY), code 125 (KEY_LEFTMETA), value 1
> Event: time 1596997549.252782, type 1 (EV_KEY), code 125 (KEY_LEFTMETA), value 0
> ```

I'm pretty sure that's not the answer he's looking for.  Sure, you are sending the right keycodes, but did you register the device as having them?  What does "evtest" show for your device config?  If your device doesn't advertise any keys, it could likely be ignored by libinput.  When you inherit from another device, those properties will get copied into yours.

e.g.
My keyboard:
```
Supported events:
  Event type 0 (EV_SYN)
  Event type 1 (EV_KEY)
    Event code 1 (KEY_ESC)
    Event code 2 (KEY_1)
    Event code 3 (KEY_2)
    [...]
```
My mouse:
```
Supported events:
  Event type 0 (EV_SYN)
  Event type 1 (EV_KEY)
    Event code 272 (BTN_LEFT)
    Event code 273 (BTN_RIGHT)
    [...]
```

**Comment by StanGenchev** [comment]
> > > Just for clarity: Did you register also all the possible KEY__, BTN__, ... codes for EV_KEY? Only having EV_KEY would probably not help libinput at all since vastly different equipment can use that.
> > 
> > 
> > Yes, it did register. Here is the output:
> > ```
> > Event: time 1596997549.252782, type 1 (EV_KEY), code 125 (KEY_LEFTMETA), value 1
> > Event: time 1596997549.252782, type 1 (EV_KEY), code 125 (KEY_LEFTMETA), value 0
> > ```
> 
> I'm pretty sure that's not the answer he's looking for. Sure, you are sending the right keycodes, but did you register the device as having them? What does "evtest" show for your device config? If your device doesn't advertise any keys, it could likely be ignored by libinput. When you inherit from another device, those properties will get copied into yours.

I have not registered anything in code, I never needed to.
All I do is this:

```
key_input = UInput()
key_input.write(ecodes.EV_KEY, ecodes.KEY_LEFTMETA, 1)
key_input.write(ecodes.EV_KEY, ecodes.KEY_LEFTMETA, 0)
```

With that code, ```evtest``` shows:

```
Supported events:
  Event type 0 (EV_SYN)
  Event type 1 (EV_KEY)
    Event code 1 (KEY_ESC)
    Event code 2 (KEY_1)
    Event code 3 (KEY_2)
    Event code 4 (KEY_3)
    .....
```
So the keys are registered or at least the default ones. I will try to find the difference between this and the one I get from create with ```from_device``` and I will report back.

**Comment by ssieb** [comment]
Right, I just checked the code and if you don't provide the list of events to use, it defaults to all keys (and buttons?).

**Comment by LinusCDE** [comment]
> I'm pretty sure that's not the answer he's looking for.

That's correct. I personally assumed that he did when showing the output. But then I also started with one of the most cumbersome devices to get started with: Graphic Tablets. Since it defaults to some Mouse/Keyboard stuff, most people don't need to bother with it usually it seems.

>After some experimentation, I found out that if I use the from_device function like this
>
>     key_input = UInput.from_device(<evdev.InputDevice>)
>
> everything works flawlessly, with any version of libinput. Before that I used
>
>     key_input = UInput()
>
> directly, as it did work, but since version 1.15.902 (and 1.16.0) of libinput, that no longer works.

Yeah. I guess libinput went stricter on what devices need to qualify for certain classes (Keyboard, Mouse, etc.). While this breaks compatibility, I can understand their reason I doing so, since the input system in linux is not just a glorified mouse & keyboard interface, but is supposed to allow any kind of device to be recognized as such by using the predefined capabilities/events.

Btw. here is the [source of from_device](https://github.com/gvalkov/python-evdev/blob/master/evdev/uinput.py#L37). All it does is passing stuff to `UInput()` but using the same events of the device you specified.

---

#### Reasons why I see importance in specifying supported events (can skip this, mostly annecdotes and personal experience).

This embedded eink notetaking device for example has all of its inputs available as input devices.
![screenshot](https://user-images.githubusercontent.com/22298664/89902853-403eb080-dbe7-11ea-8c5a-571547b5ad90.png)

That device has some buttons, wacom pen support and a typical touchscreen layer. There is no keyboard at all but all devices can be recognized as what they have by seeing their "Supported Events". The "gpio-keys" device in this example, could have any number of keys or map a "button in the middle" to whatever. Seeing all keys in advance helps a lot here.

Even the multitouch screen (mt) doesn't seem to use a common spec. Some eink reader software I ported, supported two types of they same multitouch keys that were used for Kobo ereader devices, but none of them worked for my particular device. They send some events, other touchscreens don't and behave differently. To categorize such stuff, the supported events help a lot.

Imagine adding a Joystick using just `UInput()`. Every desktop would think of it as a keyboard and/or mouse. There would be no way to know it's a joystick in advance. So it wouldn't show up in "Joystick settings" of any desktop.  
In that case product id, vendor id or product name hardly matter to a software that wants to support all kinds of joysticks.

Well, I guess that's enough rambling. Just wanted to share some experience I made with it a while ago.

**Comment by meeuw** [comment]
I have the same problem in Fedora 32 and it also automagically seems to have resolved itself.

I've noticed that libinput registers python-evdev as:
```bash
$ libinput debug-events --verbose
...
event19 - py-evdev-uinput: is tagged by udev as: Keyboard Joystick
```

And libinput might not be too happy with these Joystick capabilities. If I only add key keyboard specific events it's registered as just a keyboard.

Is there an easy way to initialize `evdev.UInput()` with only standard keyboard capabilities?

**Comment by StanGenchev** [comment]
Using libinput 1.16.1 and the issue is still present.
We need to find what changed with libinput sense v1.15.0 so that we can update the default device capabilities in python-evdev.
As I said previously, the only way to have a functioning virtual device, is to first copy the capabilities of an actual device which is not always ideal/possible.
Currently I'm berried with work, so I cannot do any research on this.

Also, sense we know that the issue is with libinput and not the kernel, should we close this issue and open another or just rename this one?

**Comment by meeuw** [comment]
Hmm, I'm still not so sure where the problem comes from (and why I can't reproduce it anymore).

It makes sense to me that libinput gets confused by python-evdev because it registers a device with so many different keys.

I've made https://github.com/gvalkov/python-evdev/pull/132 to make it easier to define an evdev with the buttons you're actually using.
